llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp

```diff
@@ -1457,13 +1457,24 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
 
     const APInt *MulC;
     if (match(Op0, m_NUWMul(m_Value(X), m_APInt(MulC)))) {
-      // Look for a "splat" mul pattern - it replicates bits across each half of
-      // a value, so a right shift is just a mask of the low bits:
-      // lshr i[2N] (mul nuw X, (2^N)+1), N --> and iN X, (2^N)-1
-      // TODO: Generalize to allow more than just half-width shifts?
-      if (BitWidth > 2 && ShAmtC * 2 == BitWidth && (*MulC - 1).isPowerOf2() &&
-          MulC->logBase2() == ShAmtC)
-        return BinaryOperator::CreateAnd(X, ConstantInt::get(Ty, *MulC - 2));
+      if (BitWidth > 2 && (*MulC - 1).isPowerOf2() &&
+          MulC->logBase2() == ShAmtC) {
+        // Look for a "splat" mul pattern - it replicates bits across each half
+        // of a value, so a right shift is just a mask of the low bits:
+        // lshr i[2N] (mul nuw X, (2^N)+1), N --> and iN X, (2^N)-1
+        if (ShAmtC * 2 == BitWidth)
+          return BinaryOperator::CreateAnd(X, ConstantInt::get(Ty, *MulC - 2));
+
+        // lshr (mul nuw (X, 2^N + 1)), N -> add nuw (X, lshr(X, N))
```

Comment:
As this transform introduces multiple uses of X, do we need a freeze here?
I don't know why alive2 doesn't complain about this.

Reply:
I'm not entirely sure, but it's possible that the nuw flag makes this safe (by converting problematic undef cases to poison).

Reply:
cc @regehr @nunoplopes 

Can you guys have a look? Thank you!

Reply:
> I'm not entirely sure, but it's possible that the nuw flag makes this safe (by converting problematic undef cases to poison).

It is confusing to me. Is this behavior documented in LangRef?
See https://alive2.llvm.org/ce/z/i2eXpk
```
; bad
define i8 @src1(i8 %x) {
  %shl = shl i8 %x, 1
  ret i8 %shl
}

define i8 @tgt1(i8 %x) {
  %add = add i8 %x, %x
  ret i8 %add
}

; good
define i8 @src2(i8 %x) {
  %shl = shl nuw i8 %x, 1
  ret i8 %shl
}

define i8 @tgt2(i8 %x) {
  %add = add i8 %x, %x
  ret i8 %add
}
```

Reply:
Nuno will have to enlighten us, I don't see why `nuw` would stop multiple uses of undef from being a problem, that doesn't make any sense to me

Reply:
`shl nuw %x, 1` implies that %x is not undef. This is because if %x is undef, we can pick a value that makes the shift yield poison.
Then we can refine poison with `add %x, %x`.

TL;DR:  `shl nuw undef, 1 -> poison -> add %x, %x`

Reply:
> `shl nuw %x, 1` implies that %x is not undef. This is because if %x is undef, we can pick a value that makes the shift yield poison. Then we can refine poison with `add %x, %x`.
> 
> TL;DR: `shl nuw undef, 1 -> poison -> add %x, %x`

It makes sense to me. Thank you!